### PR TITLE
ocean wallet grpc

### PR DESCRIPTION
**Pull request by louisinger**

### CODE REVIEW

File: internal/core/ports/wallet.go [hunk 1]
```diff
@@ -85,8 +85,8 @@ type TransactionManager interface {
 }
 
 type NotificationManager interface {
-	TxChannel() chan TxNotification
-	UtxoChannel() chan UtxoNotification
+	TxChannel() (chan TxNotification, error)
+	UtxoChannel() (chan UtxoNotification, error)
```
Comment (altafan) [inline internal/core/ports/wallet.go:89]: Better adding an `Error() error` method to both `TxNotification` and `UtxoNotification` interfaces

File: internal/core/ports/wallet.go [hunk 2]
```diff
@@ -142,7 +135,6 @@ type TxNotification interface {
 }
 
 type UtxoNotification interface {
-	AccountIndex() uint64
```
Comment (altafan) [inline internal/core/ports/wallet.go:145]: Why did you remove this?

File: cmd/tdexd/main.go [hunk 3]
```diff
@@ -70,8 +71,12 @@ func main() {
 		return
 	}
 
-	// TODO: setup a portable wallet
-	wallet, err := application.NewWallet(nil)
+	// TODO: move to config ? or constant ?
+	// who is running the ocean server ? the daemon ?
+	addrOceanServer := "localhost:50051"
```
Comment (altafan) [inline cmd/tdexd/main.go]: we will move this to config. At startup, if the daemon fails to connect to the ocean wallet, the process is stopped and the error logged to stdout.

File: internal/infrastructure/oceanwallet/ocean.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+type oceanGrpcWallet struct {
+	w ports.WalletManager
+	t ports.TransactionManager
+	n ports.NotificationManager
+	a ports.AccountManager
+}
+
+var _ ports.OceanWallet = (*oceanGrpcWallet)(nil)
```
Comment (altafan) [inline internal/infrastructure/oceanwallet/ocean.go]: what's this for??

File: internal/core/ports/wallet.go (hunk 1 above)
Comment (louisinger) [inline internal/core/ports/wallet.go:90]: Because we need to open streams connection with the gRPC server. (see [here](https://github.com/altafan/tdex-daemon/pull/1/files/0d09142cb2a0ab84003520f5beb52b37067a8acd#diff-4b210a1c4c4e9231a0edabcdad00f73182850660eafb9e109e3b420e8c6a20ffR48)) We may need a way to handle properly this "open streams" errors

File: internal/infrastructure/oceanwallet/ocean.go (hunk 4 above)
Comment (louisinger) [reply internal/infrastructure/oceanwallet/ocean.go]: it lets to ensure that `oceanGrpcWallet` implements the OceanWallet interface. If it's not the case, this line will fail at compile time

File: internal/infrastructure/oceanwallet/account_manager.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	req := &oceanv1alpha.CreateAccountRequest{
+		Name: name,
+	}
+
+	resp, err := am.client.CreateAccount(ctx, req)
+	if err != nil {
+		return 0, "", err
+	}
+
+	return resp.AccountIndex, resp.Xpub, nil
```
Comment (altafan) [inline internal/infrastructure/oceanwallet/account_manager.go]:
```suggestion
	return resp.GetAccountIndex(), resp.GetXpub(), nil
```
You should always use getters to retrieve the values from a grpc response.

This must be done basically everywhere in the oceanwallet service.

File: internal/core/ports/wallet.go (hunk 2 above)
Comment (louisinger) [reply internal/core/ports/wallet.go:145]: I think because u never use it but I can revert this. It raises a question: the `ports` interfaces should really implements ALL the proto types members ? even if we are not gonna use it inside TDEX services ?

File: internal/infrastructure/oceanwallet/wallet_manager.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	req := &oceanv1alpha.GenSeedRequest{}
+	resp, err := wm.client.GenSeed(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return strings.Split(resp.Mnemonic, " "), nil
+}
+
+func (wm *walletManagerGrpc) CreateWallet(ctx context.Context, mnemonic []string, passphrase string, chMessages chan string) (err error) {
```
Comment (altafan) [inline internal/infrastructure/oceanwallet/wallet_manager.go]:
```suggestion
func (wm *walletManagerGrpc) CreateWallet(ctx context.Context, mnemonic []string, passphrase string, chMessages chan string) error {
```

File: internal/infrastructure/oceanwallet/notifications_manager.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	txChan := make(chan ports.TxNotification)
+
+	go func() {
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				return
+			}
```
Comment (altafan) [inline internal/infrastructure/oceanwallet/notifications_manager.go]: here we break the loop and also close the channel the error is EOF, while we `continue` otherwise, maybe by debug-logging to print the error to stdout.

Comment (altafan) [reply internal/infrastructure/oceanwallet/notifications_manager.go]: Same thing applies to `UtxoChannel()`

File: internal/infrastructure/oceanwallet/notifications_manager.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	utxoChan := make(chan ports.UtxoNotification)
+
+	go func() {
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				return
+			}
+
+			utxoChan <- &utxoNotificationGrpc{resp}
```
Comment (altafan) [inline internal/infrastructure/oceanwallet/notifications_manager.go:74]: so, ideally here is where we might need to enrich the notification with block details etc for spent utxos, right?

File: internal/core/ports/wallet.go (hunk 2 above)
Comment (altafan) [reply internal/core/ports/wallet.go:145]: I will use it for settling or expring trades actually, so yes, please, revert the change.

File: internal/infrastructure/oceanwallet/ocean.go (hunk 4 above)
Comment (altafan) [reply internal/infrastructure/oceanwallet/ocean.go]: > it lets to ensure that oceanGrpcWallet implements the OceanWallet interface

This is already ensured even without your trick: the factory function has a a portable interface as return value, while the body returns the internal structure. If the body does not stick with the portable interface, the compiler complains and the compilation fails.

Comment (altafan) [reply internal/infrastructure/oceanwallet/ocean.go]: Please, remove all these lines :)

File: internal/infrastructure/oceanwallet/notifications_manager.go (hunk 8 above)
Comment (louisinger) [reply internal/infrastructure/oceanwallet/notifications_manager.go:74]: I don't think so,we need to access the `transactionManager` in order to call `GetTransaction` rpc. I think this should be done in the upper layer (application) **only when the blockdetails and txdetails are needed** : see `tradeSettleOrExpire` function as an exemple, what's your opinion ?

Comment (altafan) [reply internal/infrastructure/oceanwallet/notifications_manager.go:74]: ok fine, at app level we will eventually call the `TransactionManager().GetTransaction(...)`  API to retrieve missing data if needed.

File: internal/core/application/trade_service.go [hunk 9]
```diff
@@ -473,8 +473,11 @@ func (t *tradeService) tradeSettleOrExpire(
 	return func(notification ports.UtxoNotification) bool {
 		utxo := notification.Utxo()
 		eventType := notification.EventType()
-		txDetails := notification.TxDetails()
-		blockDetails := notification.BlockDetails()
+		txDetails, blockDetails, err := t.wallet.TransactionManager().GetTransaction(context.Background(), utxo.TxID())
+		if err != nil {
+			return false
+		}
```
Comment (altafan) [inline internal/core/application/trade_service.go:479]: this should be done only for spent unspent by prior checking the UtxoEventType
